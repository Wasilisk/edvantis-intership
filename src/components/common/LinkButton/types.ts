import {ButtonProps} from "../Button/types";

export type ButtonLinkProps = ButtonProps & {to: string};