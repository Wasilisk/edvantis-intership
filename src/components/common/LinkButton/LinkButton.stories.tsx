import {Story} from "@storybook/react";
import React from "react";
import CustomLink from "./index";
import LinkButton from "./index";
import {LinkButtonProps} from "./types";

export default {
    title: 'Application/Links/LinkButton',
    component: CustomLink,
    argTypes: {
        variant: {
            options: ['filled', 'outlined'],
            control: { type: 'radio' },
        },
        size: {
            options: ['small', 'medium', 'large'],
            control: { type: 'select' },
        },
        hoverVariant: {
            options: ['dark', 'light'],
            control: { type: 'select' },
        },
        hasArrow: {
            control: "boolean"
        },
        fullWidth: {
            control: "boolean"
        }
    }
};

const Template: Story<LinkButtonProps> = (props) => <LinkButton {...props} to="/link"/>;

export const OutlinedLinkButton = Template.bind({});
OutlinedLinkButton.args= {
    variant: "outlined",
    label: "Outlined",
    size: "medium",
    fullWidth: false
};

export const FilledLinkButton = Template.bind({});
FilledLinkButton.args= {
    variant: "filled",
    label: "Filled",
    size: "medium",
    hoverVariant: "dark",
    fullWidth: false
};

export const LinkButtonWithArrow = Template.bind({});
LinkButtonWithArrow.args= {
    variant: "filled",
    label: "Filled",
    size: "medium",
    hoverVariant: "dark",
    fullWidth: false,
    hasArrow: true
};