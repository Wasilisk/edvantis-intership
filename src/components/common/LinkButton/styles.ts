import styled from "styled-components";
import {Link} from "react-scroll";

export const LinkButtonStyles = styled(Link)`
  text-decoration: none;
`;