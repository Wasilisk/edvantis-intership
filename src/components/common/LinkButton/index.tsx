import {ButtonStyles} from "../Button/styles";
import ButtonArrow from "../Button/ButtonArrow";
import React from "react";
import {ButtonLinkProps} from "./types";
import { LinkButtonStyles } from "./styles";

const LinkButton = ({
                        to,
                        label,
                        hasArrow = false,
                        fullWidth = false,
                        hoverVariant = "dark",
                        ...props
                    }: ButtonLinkProps) => {
    return (
        <LinkButtonStyles to={to}>
            <ButtonStyles fullWidth={fullWidth} hoverVariant={hoverVariant} {...props}>
                {label}
                {
                    hasArrow && <ButtonArrow/>
                }
            </ButtonStyles>
        </LinkButtonStyles>
    );
};

export default LinkButton;