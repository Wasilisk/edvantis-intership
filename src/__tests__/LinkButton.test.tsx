import {fireEvent} from "@testing-library/react";
import LinkButton from "../components/common/LinkButton";
import {mockHistory} from "./__utils__/history-wrapper";
import { render } from "./__utils__/store-wrapper";

describe('LinkButton component', () => {
    it('should render outlined button', () => {
        const {asFragment} = render(<LinkButton
            variant="outlined"
            size="small"
            label="Click me"
            to="/link"

        />);
        expect(asFragment()).toMatchSnapshot();
    })
    it('should render filled button with dark hover', () => {
        const {asFragment} = render(<LinkButton
            variant="filled"
            size="small"
            label="Click me"
            to="/link"
        />);
        expect(asFragment()).toMatchSnapshot();
    })
    it('should render filled button with light hover', () => {
        const {asFragment} = render(<LinkButton
            variant="filled"
            size="small"
            label="Click me"
            hoverVariant="light"
            to="/link"
        />);
        expect(asFragment()).toMatchSnapshot();
    })
    it('should render fullwidth button', () => {
        const {asFragment} = render(<LinkButton
            variant="outlined"
            size="medium"
            label="Click me"
            fullWidth
            to="/link"
        />);
        expect(asFragment()).toMatchSnapshot();
    })
    it('should render button with arrow', () => {
        const {container, asFragment} = render(<LinkButton
            variant="outlined"
            size="medium"
            label="Click me"
            fullWidth
            hasArrow
            to="/link"
        />);

        const buttonArrow = container.querySelector('button');
        expect(buttonArrow).toBeDefined()
        expect(asFragment()).toMatchSnapshot();
    })
    it('should redirect when clicked', () => {
        const {container, asFragment} = render(<LinkButton
            variant="outlined"
            size="medium"
            label="Click me"
            to="/link"
        />);

        fireEvent.click(container.firstChild!);
        expect(mockHistory.location.pathname).toBe("/link");

        expect(asFragment()).toMatchSnapshot();
    })
})