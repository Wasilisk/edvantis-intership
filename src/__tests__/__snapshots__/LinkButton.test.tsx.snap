// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkButton component should redirect when clicked 1`] = `
<DocumentFragment>
  <a
    class="sc-gsnTZi dPktQh"
    href="/link"
  >
    <button
      class="sc-bczRLJ fnaGjx"
    >
      Click me
    </button>
  </a>
</DocumentFragment>
`;

exports[`LinkButton component should render button with arrow 1`] = `
<DocumentFragment>
  <a
    class="sc-gsnTZi dPktQh"
    href="/link"
  >
    <button
      class="sc-bczRLJ fAvsjm"
    >
      Click me
      <svg
        height="100%"
        viewBox="0 0 36 36"
        width="100%"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <path
          class="cls-1"
          d="M34,18l-5-4v3H26v2h3v3ZM24,19H20V17h4v2Zm-6,0H14V17h4v2Zm-6,0H8V17h4v2ZM4,16a2,2,0,1,1-2,2A2,2,0,0,1,4,16Z"
          data-name="28"
          fill="currentColor"
        />
      </svg>
    </button>
  </a>
</DocumentFragment>
`;

exports[`LinkButton component should render filled button with dark hover 1`] = `
<DocumentFragment>
  <a
    class="sc-gsnTZi dPktQh"
    href="/link"
  >
    <button
      class="sc-bczRLJ jJNmHb"
    >
      Click me
    </button>
  </a>
</DocumentFragment>
`;

exports[`LinkButton component should render filled button with light hover 1`] = `
<DocumentFragment>
  <a
    class="sc-gsnTZi dPktQh"
    href="/link"
  >
    <button
      class="sc-bczRLJ dhZYZA"
    >
      Click me
    </button>
  </a>
</DocumentFragment>
`;

exports[`LinkButton component should render fullwidth button 1`] = `
<DocumentFragment>
  <a
    class="sc-gsnTZi dPktQh"
    href="/link"
  >
    <button
      class="sc-bczRLJ fAvsjm"
    >
      Click me
    </button>
  </a>
</DocumentFragment>
`;

exports[`LinkButton component should render outlined button 1`] = `
<DocumentFragment>
  <a
    class="sc-gsnTZi dPktQh"
    href="/link"
  >
    <button
      class="sc-bczRLJ hdOqBX"
    >
      Click me
    </button>
  </a>
</DocumentFragment>
`;
